cache/disk/disk.go

```diff
@@ -288,161 +242,220 @@ func (c *Cache) Put(kind cache.EntryKind, hash string, expectedSize int64, r io.
+	tf, err := tfc.Create(filePath)
+	if err != nil {
+		return err
+	}
+	removeTempfile = true
 
-	// Download to a temporary file
-	tmpFilePath := filePath + ".tmp"
-	f, err := os.Create(tmpFilePath)
+	err = writeAndCloseFile(r, kind, hash, size, tf)
 	if err != nil {
 		return err
 	}
+
+	if c.proxy != nil {
+		rc, err := os.Open(tf.Name())
+		if err != nil {
+			log.Println("Failed to proxy Put:", err)
+		} else {
+			// Doesn't block, should be fast.
+			c.proxy.Put(kind, hash, size, rc)
+		}
+	}
+
+	unreserve, removeTempfile, err = c.commit(key, tf.Name(), filePath, size, size)
+
+	// Might be nil.
+	return err
+}
+
+func writeAndCloseFile(r io.Reader, kind cache.EntryKind, hash string, size int64, f *os.File) error {
+	closeFile := true
 	defer func() {
-		if !shouldCommit {
-			// Only delete the temp file if moving it didn't succeed.
-			os.Remove(tmpFilePath)
+		if closeFile {
+			f.Close()
 		}
-		// Just in case we didn't already close it.  No need to check errors.
-		f.Close()
 	}()
 
+	var err error
+
 	var bytesCopied int64
 	if kind == cache.CAS {
 		hasher := sha256.New()
-		if bytesCopied, err = io.Copy(io.MultiWriter(f, hasher), r); err != nil {
+		bytesCopied, err = io.Copy(io.MultiWriter(f, hasher), r)
+		if err != nil {
 			return err
 		}
 		actualHash := hex.EncodeToString(hasher.Sum(nil))
 		if actualHash != hash {
 			return fmt.Errorf(
-				"hashsums don't match. Expected %s, found %s", key, actualHash)
+				"checksums don't match. Expected %s, found %s", hash, actualHash)
 		}
 	} else {
 		if bytesCopied, err = io.Copy(f, r); err != nil {
 			return err
 		}
 	}
 
+	if isSizeMismatch(bytesCopied, size) {
+		return fmt.Errorf(
+			"sizes don't match. Expected %d, found %d", size, bytesCopied)
+	}
+
 	if err = f.Sync(); err != nil {
 		return err
 	}
 
 	if err = f.Close(); err != nil {
 		return err
 	}
+	closeFile = false
 
-	if bytesCopied != expectedSize {
-		return fmt.Errorf(
-			"sizes don't match. Expected %d, found %d", expectedSize, bytesCopied)
+	return nil
+}
+
+// This must be called when the lock is not held.
+func (c *Cache) commit(key string, tempfile string, finalPath string, reservedSize int64, foundSize int64) (unreserve bool, removeTempfile bool, err error) {
+	unreserve = reservedSize > 0
+	removeTempfile = true
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if unreserve {
+		err = c.lru.Unreserve(reservedSize)
+		if err != nil {
+			log.Println(err.Error())
+			return true, removeTempfile, err
+		}
 	}
+	unreserve = false
 
-	// Rename to the final path
-	err = os.Rename(tmpFilePath, filePath)
-	if err == nil {
-		// Only commit if renaming succeeded.
-		// This flag is used by the defer() block above.
-		shouldCommit = true
+	if !c.lru.Add(key, &lruItem{size: foundSize}) {
+		err = fmt.Errorf("INTERNAL ERROR: failed to add: %s, size %d", key, foundSize)
+		log.Println(err.Error())
+		return unreserve, removeTempfile, err
 	}
 
-	return err
+	err = os.Rename(tempfile, finalPath)
```

Comment:
If the cache item filenames contain the size, then it would be difficult to service requests (like HTTP GET/HEAD) that do not specify a size.

I think we need the file renaming and insertion into the index to be atomic for a given key. Otherwise overlapping Put calls can leave the file on disk out of sync with the index.

Aside: I just notice that `os.Rename()` is not atomic, it calls `os.Lstat()` and then `syscall.Rename()`. We may as well just call `syscall.Rename()` directly.

Reply:
Hmm, another idea... what if we write to a tempfile in the dir, then rename it to something which indicates that we have finished writing to it, then obtain the lock, and lru.Add an item to the LRU which contains the final filename and the same key as now. lru.Add could then return the previous filename if there was one. Then we release the lock and remove the old file that was returned (if there was one).

On startup, if there are multiple "finished" files for a given key we keep the one with the most recent atime and delete the others.

But I think this change might be best done in a followup PR, if calling syscall.Rename turns out to be too slow.

Reply:
> If the cache item filenames contain the size, then it would be difficult to service requests (like HTTP GET/HEAD) that do not specify a size.

For PUT we know the size after writeAndCloseFile finished for the temporary file. And for GET/HEAD with unspecified size, perhaps we could get the size from LRU, via the hash as key?

> I think we need the file renaming and insertion into the index to be atomic for a given key. Otherwise overlapping Put calls can leave the file on disk out of sync with the index.

I’m thinking about if embedding both hash and size in the final file name, could avoid the risk for out of sync between file on disk and index.

> Aside: I just noticed that os.Rename() is not atomic, it calls os.Lstat() and then syscall.Rename(). We may as well just call syscall.Rename() directly.

Good finding!